### ISSUE TITLE
Skip() inside test and before block

### ISSUE THREAD

**Issue by karlkfi** [onsi/ginkgo#168]
Looking for a way to skip a test at runtime (add to pending list) based on some conditional.

Intuitively I was looking for:

```
It("does stuff", func() {
  if !supportedBy(platform) {
    Skipf("Not supported by platform %s", platform)
  }

  ExpectStuff()
}
```

I found `GinkgoT().Skipf()` put it doesn't do what I expect it to, which is exit the test (non-failure) and increment the "pending" counter printed at the end.

Could be a feature request, unless there's an idomatic way already.

**Comment by onsi** [comment]
hey @karlkfi 

There are, implicitly, two execution times in Ginkgo: tree-construction time followed by test-execution time.  Tree-construction time occurs when Go starts up the test binary and all your `Describe`s and `It`s are run.  This builds an in-memory tree of tests (closures, really).  Ginkgo then walks this tree to execute your tests.

Whether a test is marked as pending/focused/etc... must happen at tree-construction time.  It might be possible to change this but it's going to be a pain and will bend the codebase in awkward directions.

Typically, things like platform checks don't depend on any of your test code to have run and can be performed independently at tree-construction time.  So you could (for example, there are other ways to package this up):

```
func OnSupportedPlatformsIt(description string, f interface{}) {
      if supportedBy(platform) {
          It(description, f)
      } else {
          PIt(description, f)
      }
}
```

Then, in your code:

```
var _ = Describe("sprockets", func() {
    BeforeEach(func() {
        //setup
    })

    OnSupportedPlatformsIt("does something platform-specific", func() {
        ExpectStuff()
    })
})
```

If you don't care about the "pending" counter you can just `return` out of the `It`.  The only way to modify the "pending" counter is to do so at tree-construction time.

**Comment by karlkfi** [comment]
Yeah, that's kind of what I was afraid of. I think in the case I was looking at the platform was being determined at test execution time in a BeforeSuite. The options seemed to be either fail and require user to --skip or return and pass with no notification to the user other than a log hidden in the verbose output. I guess alternatively I can spawn test sub-commands, but the thought of that makes me cringe...

Would it make sense to just add a distinct "skipped counter" to display in the end summary? 
It could also have verbose output that mirrored pending output.
That would bypass having to rewrite how pending works.

**Comment by karlkfi** [comment]
Another option might be a Skip that raises a singleton panic that indicates a state other than failure...

**Comment by robdimsdale** [comment]
@onsi should this issue be closed? The corresponding PR #169 was merged.

**Comment by onsi** [comment]
Yessir, it should!
